### DISCUSSION TITLE
Viewer interaction

### DISCUSSION THREAD

**Discussion by lypwig** [bernhard-42/vscode-ocp-cad-viewer#133] [Show and tell]
Following the discussion in #126, I wanted to play with viewer interactivity today.

## Demo

The example below is not really useful but it's a fun proof of concept: it creates a smaller box on the selected face.

![output](https://github.com/user-attachments/assets/13c8613b-d7fc-4570-a8aa-262b103db49c)

## How

The base principle is inherited from my project [bd_shape_history](https://codeberg.org/roipoussiere/bd_shape_history) and consists in indexing shapes with their hash. Each hash is based on shape tessellation in order to compare them in different program executions (you can see them in the tree view above). Then, using the listener function from the ocp viewer backend, I can handle clicks on the object, which gives me the shape label, which is the shape hash, which is used to get the actual shape. I can then modify the object and update the viewer. Note that what I pass in the viewer function is not the object but a compound containing all the object faces.

## Source

```py
import json

import build123d as _
from ocp_tessellate import convert
import ocp_vscode as ov

DEFAULT_FACE_COLOR = "orange"


class PartManager:
    def __init__(self, part: _.Part, label='') -> None:
        self.part = part
        self.label = label if label else part.label

        self.part_id = self.get_shape_id(part)
        self.faces = {self.get_shape_id(f): f for f in part.faces()}
        self.faces_id, self.faces_obj = zip(*self.faces.items())

        for face_id, face in self.faces.items():
            face.label = face_id
            face.color = DEFAULT_FACE_COLOR

        self.object = _.Compound(label=self.label, children=self.faces_obj)

    def update_view(self) -> None:
        ov.show_object(self.object, clear=True)

    def get_shape_id(self, shape: _.Shape) -> str:
        subshapes = convert.tessellate_group(*convert.to_ocpgroup(shape))[1]
        part_hash = hash(json.dumps(convert.numpy_to_buffer_json(subshapes)))
        return hex(part_hash)[2:]

    def on_shape_clicked(self, shape_id: str) -> None:
        face = pm.faces[shape_id]
        box_size = face.length/2, face.width/2, (face.length + face.width)/4
        box_align = (_.Align.CENTER, _.Align.CENTER, _.Align.MIN)
        box = _.Box(*box_size, align=box_align).move(face.center_location)
        self.__init__(self.part + box, self.label)
        self.update_view()


def init_viewer(pm: PartManager) -> None:
    from ocp_vscode.comms import listener, MessageType

    def handle_event(message, event_type: MessageType) -> None:
        if event_type == MessageType.UPDATES:
            if "selectedShapeIDs" in message:
                selected = message["selectedShapeIDs"]
                for shape_id in [sid.split('/')[2] for sid in selected]:
                    pm.on_shape_clicked(shape_id)

    listener(handle_event)()


pm = PartManager(_.Box(4, 3, 2), 'boite')
init_viewer(pm) # comment this line to initialize the viewer again
pm.update_view()
```

**Comment by bernhard-42** [comment]
nice!
Glad to see that the python listener found another use.

What is your overall goal?

**Reply by lypwig** [reply]
I don't have specific goal but I'm pretty sure that the ability to execute custom code based on selected face could improve the way users create things with cadquery / build123d and will allow the emergence of many tools used for debugging or code generation.

For instance, in my project [bd_shape_history](https://codeberg.org/roipoussiere/bd_shape_history), this could be used to understand which line of code is responsible of the creation of the selected face.
